src/createApp.ts

```diff
@@ -837,6 +852,48 @@ const createApp = compose({
 			return instanceRegistries.get(this).identifyStore(store);
 		},
 
+		createWidget<U extends Child, O>(this: App, factory: ComposeFactory<U, O>, options: any = {}): Promise<[ string, U ]> {
+			const app = this;
+			const { defaultWidgetStore, registryProvider } = this;
+			const { id = generateWidgetId() } = options;
+			// Like for custom elements, don't add the generated ID to the options.
+
+			function configureWidget(): Promise<any> | void {
+				// Ensure no other widget with this ID exists.
+				if (publicRegistries.get(app).hasWidget(id)) {
+					return Promise.reject(new Error(`A widget with ID '${id}' already exists`));
+				}
+
+				if (!options.registryProvider) {
+					options.registryProvider = registryProvider;
+				}
+
+				if (options.id && !options.stateFrom && defaultWidgetStore) {
```

Comment:
I think we always want widgets to have an entry in the widget store and the parent to have a reference to the array of children. It will probably be easier to reason about with consistent behaviour, we can always enhance this in the future to add the optional path later?